File: scripts/adjust_gday_param_file.py
#!/usr/bin/env python

"""
Change various params in a G'DAY input file.

All that needs to be supplied is a dictionary with the variable to be changed
and the corresponding paramter value (oh and of course the param fname!).
"""

import os
import re
import tempfile
import shutil

__author__  = "Martin De Kauwe"
__version__ = "1.0 (09.06.2011)"
__email__   = "mdekauwe@gmail.com"

def adjust_param_file(fname, replacements):
    """ adjust model parameters in the file and save over the original.

    Parameters:
    ----------
    fname : string
        parameter filename to be changed.
    replacements : dictionary
        dictionary of replacement values.

    """
    fin = open(fname, 'r')
    param_str = fin.read()
    fin.close()
    new_str = replace_keys(param_str, replacements)
    fd, path = tempfile.mkstemp()
    os.write(fd, new_str)
    os.close(fd)
    shutil.copy(path, fname)
    os.remove(path)

def replace_keys(text, replacements_dict):
    """ Function expects to find GDAY input file formatted key = value.

    Parameters:
    ----------
    text : string
        input file data.
    replacements_dict : dictionary
        dictionary of replacement values.

    Returns:
    --------
    new_text : string
        input file with replacement values

    """
    lines = text.splitlines()
    for i, row in enumerate(lines):
        # skip blank lines
        if not row.strip():
            continue
        # skip .cfg section dividers
        elif not row.startswith("["):
            key, sep, val = row.split()
            lines[i] = " ".join((key, sep, replacements_dict.get(key, val)))
        elif row.startswith("[print]"):
            break


    return '\n'.join(lines) + '\n'


if __name__ == "__main__":

    cfg_fname = "/Users/mdekauwe/research/NCEAS_face/GDAY_duke_simulation/params/dk_varyco2_varyndep_grassequilib_then_forest_dukegrass_youngforest.cfg"
    replace_dict = { "albedo": "0.123"}
    adjust_param_file(cfg_fname, replace_dict)
    print cfg_fname

File: scripts/adjust_gday_param_file.R
# Change various params in a G'DAY input file.
# Author: Martin De Kauwe
# Date: 12.05.2016
# Email: mdekauwe@gmail.com
#
# e.g....
#
# in_fname <- "base_start.cfg"
# out_fname <- "test.ini"
#
# replacements <- list("ncycle" = "true",
#                     "modeljm" = "3",
#                     "print_options" = "end",
#                     "jmax" = "110.0",
#                     "vcmax" = "55.0")
#
# adjust_gday_params(in_fname, out_fname, replacements)

adjust_gday_params <- function(in_fname, out_fname, replacements) {

  if (!require("ini")){
      install.packages("ini")
      library(ini)
  }

  g <- read.ini(in_fname)

  for (key in names(replacements)) {

    match_git <- key %in% names(g$git)
    match_files <- key %in% names(g$files)
    match_params <- key %in% names(g$params)
    match_state <- key %in% names(g$state)
    match_control <- key %in% names(g$control)

    if (match_git) {
      g$git[key] <- replacements[key]
    } else if (match_files) {
      g$files[key] <- replacements[key]
    } else if (match_params) {
      g$params[key] <- replacements[key]
    } else if (match_state) {
      g$state[key] <- replacements[key]
    } else if (match_control) {
      g$control[key] <- replacements[key]
    }

  }
  write.ini(g, out_fname)
}
